### PR TITLE
Basic implementation for filesystems table

### PR DESCRIPTION
**Pull request by zwass**
Example results
```
osquery> select * from filesystems;

+---------------+-------------------+--------+
| name          | path              | type   |
+---------------+-------------------+--------+
| /dev/disk1    | /                 | hfs    |
| devfs         | /dev              | devfs  |
| map -hosts    | /net              | autofs |
| map auto_home | /home             | autofs |
| map -fstab    | /Network/Servers  | autofs |
| /dev/disk3    | /Volumes/RAM Disk | hfs    |
| /dev/disk4    | /Volumes/woooop   | hfs    |
+---------------+-------------------+--------+
```

Note that there is no indication that a ram disk is actually a ram disk. Does the OS know that?

A couple questions here:

1) How should I test this? Abstract most of genFilesystems into another function in a filesystems.h and "unit" test that some results are returned? Actually mock out the call to getfsstat?

2) What other data do we want to try to retrieve? Available data can be seen in the "description" section of the [statfs man page](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/statfs.2.html#//apple_ref/doc/man/2/statfs). Note that in tests on my local Mac `f_owner`, `f_fssubtype` and `f_flags` (the other interesting seeming fields) all had `-1` values to indicate that they were undefined.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the admins verify this patch? Reply with "ok to test" to kick off a build...

**Comment by marpaia** [comment]
ok to test

**Comment by marpaia** [comment]
This looks great. With regards to testing, I usually do something like this: https://github.com/facebook/osquery/blob/master/osquery/tables/system/darwin/firewall.cpp#L215-L222

Abstract out the systems interactions into very small isolated spots and then test the data manipulation and parsing with some dummy data: https://github.com/facebook/osquery/blob/master/osquery/tables/system/darwin/firewall_tests.cpp

With that being said, a nice mock setup would be much more sustainable and correct.

File: osquery/tables/system/darwin/filesystems.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    // First get count of filesystems
+    int fs_count = getfsstat(NULL, 0, 0);
+    if (fs_count == -1) {
+      LOG(ERROR) << "Error retrieving filesystems count from getfsstat";
+      return {};
+    }
+
+    size_t fs_infos_size = fs_count * sizeof(struct statfs);
+    std::unique_ptr<struct statfs[]> fs_infos(new struct statfs[fs_infos_size]);
```
Comment (marpaia) [inline osquery/tables/system/darwin/filesystems.cpp]: image

File: osquery/tables/system/darwin/filesystems.cpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#include <sys/mount.h>
+
+#include <glog/logging.h>
+
+#include <osquery/tables.h>
+
+namespace osquery {
+namespace tables {
+
+  QueryData genFilesystems(QueryContext& context) {
```
Comment (marpaia) [inline osquery/tables/system/darwin/filesystems.cpp]: We usually don't indent the namespace body, but clang-format will take care of things like that.

**Comment by theopolis** [comment]
Wait, I'm confused, how is this different than the existing mounts table?

```python
table_name("mounts")
schema([
	Column("device", TEXT),
	Column("device_alias", TEXT),
	Column("path", TEXT),
	Column("type", TEXT),
	Column("blocks_size", BIGINT),
	Column("blocks", BIGINT),
	Column("blocks_free", BIGINT),
	Column("blocks_available", BIGINT),
	Column("inodes", BIGINT),
	Column("inodes_free", BIGINT),
	Column("flags", TEXT),
])
implementation("mounts@genMounts")
```

```
+--------------------------------------------------------+--------------+----------+----------+-------------+---------+-------------+------------------+---------+-------------+-------------------------------------------------------+
| device                                                 | device_alias | path     | type     | blocks_size | blocks  | blocks_free | blocks_available | inodes  | inodes_free | flags                                                 |
+--------------------------------------------------------+--------------+----------+----------+-------------+---------+-------------+------------------+---------+-------------+-------------------------------------------------------+
| rootfs                                                 | rootfs       | /        | rootfs   | 4096        | 4869551 | 3545909     | 3292789          | 1245184 | 1080003     | rw                                                    |
| sysfs                                                  | sysfs        | /sys     | sysfs    | 4096        | 0       | 0           | 0                | 0       | 0           | rw,nosuid,nodev,noexec,relatime                       |
| proc                                                   | proc         | /proc    | proc     | 4096        | 0       | 0           | 0                | 0       | 0           | rw,nosuid,nodev,noexec,relatime                       |
| udev                                                   | udev         | /dev     | devtmpfs | 4096        | 251711  | 251710      | 251710           | 251711  | 251235      | rw,relatime,size=1006844k,nr_inodes=251711,mode=755   |
| devpts                                                 | devpts       | /dev/pts | devpts   | 4096        | 0       | 0           | 0                | 0       | 0           | rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 |
| tmpfs                                                  | tmpfs        | /run     | tmpfs    | 4096        | 50883   | 50714       | 50714            | 254413  | 253954      | rw,nosuid,noexec,relatime,size=203532k,mode=755       |
| /dev/disk/by-uuid/f014d386-fa05-4e8a-b23f-1a112806d082 | /dev/sda1    | /        | ext4     | 4096        | 4869551 | 3545909     | 3292789          | 1245184 | 1080003     | rw,relatime,errors=remount-ro,data=ordered            |
+--------------------------------------------------------+--------------+----------+----------+-------------+---------+-------------+------------------+---------+-------------+-------------------------------------------------------+
```

Sorry I haven't commented/reviewed sooner. @marpaia  and @zwass I saw some discussion about a filesystems table before and wasn't sure what was recommended.

**Comment by zwass** [comment]
Ah yeah this looks to be exactly the same as that. I was working off of #13 which is maybe just an irrelevant task now?

**Comment by zwass** [comment]
This was looking at the data from the same struct as the existing mounts table. It looks like it should probably just be abandoned.

**Comment by marpaia** [comment]
https://31.media.tumblr.com/3dcc8f3502cadf664d4308eabc83798f/tumblr_n9qzt4c3TT1rbc9h1o1_500.gif

**Comment by zwass** [comment]
I noted on the issue, and closed.

**Comment by marpaia** [comment]
I don't think it's worth closing this just yet. The mounts table has two implementation files (system/linux/mounts.cpp and system/darwin/mounts.cpp). Is this implementation more cross platform?

**Comment by osqueryer** [comment]
Can one of the admins verify this patch? Reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the admins verify this patch? Reply with "ok to test" to kick off a build...

**Comment by marpaia** [comment]
be quiet, @osqueryer

**Comment by zwass** [comment]
Unfortunately I don't think this is really more cross platform. The [`getfsstat`](http://manpages.ubuntu.com/manpages/hardy/man2/getfsstat.2.html) syscall exists on Ubuntu, but the [`struct statfs`](http://manpages.ubuntu.com/manpages/hardy/man2/statfs.2.html) seems to be different there. Potentially some of the implementation could be shared, but a major difference is that the Ubuntu call doesn't fill in any strings.

**Comment by theopolis** [comment]
Cleaning up again :dancers:

**Pull request closed without merging**